lib/services/suppressions-service.js

```diff
@@ -0,0 +1,282 @@
+
+                // Cleanup files with no rules
+                delete suppressions[file];
+            }
+        }
+
+        return this.save(suppressions);
+    }
+
+    /**
+     * Checks the provided suppressions against the lint results.
+     *
+     * For each file, counts the number of violations per rule.
+     * For each rule in each file, compares the number of violations against the counter from the suppressions file.
+     * If the number of violations is less or equal to the counter, messages are moved to `LintResult#suppressedMessages` and ignored.
+     * Otherwise, all violations are reported as usual.
+     * @param {LintResult[]} results The lint results.
+     * @param {SuppressedViolations} suppressions The suppressions.
+     * @returns {{
+     *   results: LintResult[],
+     *   unused: SuppressedViolations
+     * }} The updated results and the unused suppressions.
+     */
+    applySuppressions(results, suppressions) {
+
+        /**
+         * We copy both the results and the suppressions to avoid modifying the original objects
+         * We remove only suppressions and result messages that are matched and hence suppressed
+         * We leave the rest untouched to minimize the risk of losing parts of the original data
+         */
+        const filtered = structuredClone(results);
+        const unused = structuredClone(suppressions);
+
+        for (const result of filtered) {
+            const relativeFilePath = this.getRelativeFilePath(result.filePath);
+
+            if (!suppressions[relativeFilePath]) {
+                continue;
+            }
+
+            const violationsByRule = SuppressionsService.countViolationsByRule(result.messages);
+
+            for (const ruleId in violationsByRule) {
+                if (!unused[relativeFilePath][ruleId]) {
+                    continue;
+                }
+
+                const suppressionsCount = unused[relativeFilePath][ruleId].count;
+                const violationsCount = violationsByRule[ruleId].count;
+
+                // Suppress messages if the number of violations is less or equal to the suppressions count
+                if (violationsCount <= suppressionsCount) {
+                    SuppressionsService.suppressMessagesByRule(result, ruleId);
+                }
+
+                // Update the count to match the new number of violations, otherwise remove the rule entirely
+                if (violationsCount < suppressionsCount) {
+                    unused[relativeFilePath][ruleId].count = suppressionsCount - violationsCount;
+                } else {
+                    delete unused[relativeFilePath][ruleId];
+                }
+            }
+
+            // Cleanup files with no rules
+            if (Object.keys(unused[relativeFilePath]).length === 0) {
+                delete unused[relativeFilePath];
+            }
+        }
+
+        return {
+            results: filtered,
+            unused
```

Comment:
This returns suppressions for files that aren't linted as unused.

For example, if eslint was first run on two files, with `--suppress-all`:

```
eslint foo.js bar.js --suppress-all
```

it will store entries for `foo.js` and `bar.js` in `.eslint-suppressions.json`.

Then, if eslint is run on only one file:

```
eslint foo.js
```

it will find the entry for `bar.js` as unused, and report an error, with exit code 2:

```
There are suppressions left that do not occur anymore. Consider re-running the command with `--prune-suppressions`.
```

I think only suppressions for files that are included in linting should be checked. Thoughts?

Reply:
I agree, good point. The implementation of Suppressions service assumes that eslint runs on the same files every time. I suggest to change the implementation so that Suppressions service is aware of the `files` from `cli.js`.

In particular:

* `--suppress-all` will generate suppressions only for the files that are included in linting. This means that the suppressions file will never get over-written. 
* `--suppress-rule` will also generate suppressions only for the files that are included in linting.
* `--prune-suppressions` will also prune suppressions only for the files that are included in linting.
* As suggested, the unused suppressions will be reported only for files that are included in linting.
* Adjust docs to reflect the above changes

Let me know what you think.

Reply:
Having a closer look at the code, it seems that relying only on `files` won't be sufficient due to the support for globs and others. `findFiles` helper can help but I would like to avoid execute the process twice. We could explore using `findFiles` directly from `cli.js`, but that would bring backwards incompatible changes to ESLint class

Reply:
Files that were included in linting are those present in `results` passed to this method.